Override and block policies for libraries (#3526)
chainguard-dev/edu: content/chainguard/libraries/access.md

## Manage library policies

> Users with the Owner role can create, enable, disable, and list library policies using [`chainctl libraries policy`](/chainguard/chainctl/chainctl-docs/chainctl_libraries_policy/) commands. These policies apply to upstream packages pulled through Chainguard Repository, so [upstream fallback](/chainguard/libraries/overview/#upstream-fallback-and-controls) must be enabled for an ecosystem before you apply a rule to it. 

Reviewer: The cooldown policy applies to all packages pulled through Chainguard Repository (not just upstream ones)

Today you have to have upstream enabled to create a cooldown policy, I'm actually not sure what the behavior is for custom package blocks though. We'll need to test this + confirm with eng

Only one custom policy per ecosystem can be enforced at a time, so include all the rules you need - cooldown, block, and override - in the same policy for that ecosystem.

Chainguard Libraries supports the following types of policies:

> - **Cooldown**: Delays newly published package versions for a set number of days, giving the broader security community time to surface threats that malware and greyware scanning may not catch immediately.

Reviewer: should we reword or remove the 2nd half of this given Manfred's feedback?

> - **Cooldown**: Delays newly published package versions for a set number of days, giving the broader security community time to surface threats that malware and greyware scanning may not catch immediately.

Reviewer 2: Can we clarify what "newly published" means? I assume we are talking about the publishing date in the upstream repository and NOT when we ended up building a library in our own repo. The reason I am asking is that @andrewd-cg mentioned that cooldown also applies to the libraries we built, but it is not clear to me what date matters there. For example, in the console, we weirdly show the date a library was built by us .. and not when it was built/published by the upstream project

Author: @angela-zhang  does the cooldown timeframe refer only to the upstream published date?

Reviewer: yep, it's based on the upstream publish date

- **Block**: Deny a package or version outright. Use a block rule when your organization never wants to allow a package or version.

> - **Override**: Permit a package or version that would otherwise be denied by cooldown or malware gating. Use an override when you need a specific, deliberate exception.

Reviewer: ```suggestion
- **Override**: Permit a package or version that would otherwise be denied by a cooldown policy or malware and greyware blocking. Use an override when you need a specific, deliberate exception.
```

> - **Override**: Permit a package or version that would otherwise be denied by cooldown or malware gating. Use an override when you need a specific, deliberate exception.

Reviewer 2: Can an override also be used against a block?

Reviewer: the override allowlist will always take precedence.

Reviewer 2: Okay .. we should make sure this is documented

>Note: The commands in this section require `chainctl` v0.2.291 or newer.

#### Identify packages with a purl

> Library policies use package URLs, or purls, to identify packages across ecosystems:

Reviewer 2: Might be worth linking to https://spdx.github.io/spdx-spec/v3.0.1/annexes/pkg-url-specification/

> - Python: `pkg:pypi/<name>`

Reviewer 2: Should add more info here also about version and have actual real world examples.

- JavaScript: `pkg:npm/<name>` or `pkg:npm/%40<scope>/<name>` for scoped packages
- Java: `pkg:maven/<group>/<artifact>`

> Append `@<version>` to target one version. Omit the version to match all versions of the package

Reviewer: ```suggestion
Append `@<version>` to target one version. Omit the version to include all versions of the package.
```

### Preview a policy

> To understand the impact before enforcing a policy, use `--mode=PREVIEW`. In preview mode, installs continue to succeed, but Chainguard records what _would have been blocked_ if the policy were enforced.

Reviewer 2: Can you explain this with some examples? I am not clear on how this works with multiple purl blocks, overrides, and such. Also, what is the reference set of libraries it looks against for the policy? 

But I now see this is coming later..

Author: I added more details here

For example:

```bash

> chainctl libraries policy create --name=disable-cooldown --cooldown-days=0

Reviewer: i think this example should be a differenet number of cooldown days (a 0 day cooldown wouldn't block anything)

## Manage library policies

> Users with the Owner role can create, enable, disable, and list library policies using [`chainctl libraries policy`](/chainguard/chainctl/chainctl-docs/chainctl_libraries_policy/) commands. These policies apply to all packages pulled through Chainguard Repository; [upstream fallback](/chainguard/libraries/overview/#upstream-fallback-and-controls) must be enabled for an ecosystem before you apply a rule to it. 

Reviewer: ```suggestion
Users with the Owner role can create, enable, disable, and list library policies using [`chainctl libraries policy`](/chainguard/chainctl/chainctl-docs/chainctl_libraries_policy/) commands. These policies apply to all packages pulled through Chainguard Repository. The [upstream fallback](/chainguard/libraries/overview/#upstream-fallback-and-controls) must be enabled for an ecosystem in order to use policies.
```

> Only one custom policy per ecosystem can be enforced at a time, so include all the rules you need - cooldown, block, and override - in the same policy for that ecosystem.

Reviewer: ```suggestion
One custom policy per ecosystem can be enabled at a time. This policy should include all the rules you need, which may include a cooldown period, package block rules, and any overrides.
```

Chainguard Libraries supports the following types of policies:

> - **Cooldown**: Delays upstream newly published package versions for a set number of days.

Reviewer: ```suggestion
- **Cooldown**: Delays newly published package versions for a set number of days, after the upstream registry publish date.
```

chainctl libraries policy enable --policy=no-cooldown --ecosystem=JAVA --mode=ENFORCE
```

### Block a package or version

Create a custom policy with one or more `--block` entries to deny packages explicitly. For example:

```bash
chainctl libraries policy create --name=team-policy \

>   --block=purl=pkg:npm/left-pad

Reviewer: can we include an example with multiple packages on the blocklist?

Reviewer 2: Yes and also .. how would a team verify what actually breaks then. This is a good example, leftpad is so commonly used that there is a high chance that a project would no longer build since leftpad comes in as a transitive dependency. It might be better to have an example of a high level package that is not allowed in an org .. like maybe Angular is not allowed because some other framework is the standard..

Author: I updated to include multiple packages and gave an example of standardizing on React and blocking Angular packages

chainctl libraries policy enable --name=team-policy --ecosystem=JAVASCRIPT --mode=ENFORCE
```

To block one specific version, include the version in the purl. For example:

```bash
chainctl libraries policy create --name=team-policy \

>   --block=purl=pkg:npm/lodash@4.17.20

Reviewer 2: Good specific example but we should have a reason to block this that we mention here

Author: I found a recent CVE example and described blocking the last affected version, but maybe this isn't a good example unless the policy lists out blocks for all the affected versions?

Reviewer 2: Not sure tbh

Author: I ended up removing the specifics of the example, and used a package/version with a known issue

chainctl libraries policy enable --name=team-policy --ecosystem=JAVASCRIPT --mode=ENFORCE
```

To block all versions of a package, omit the version. For example:

```bash

> chainctl libraries policy update team-policy \

Reviewer: I think we need to clarify that a current limitation is that this will replace the existing blocklist (so customers must re-iterate the entire existing blocklist including anything new they want to block). we are working to improve this but should document for now

  --block=purl=pkg:pypi/<name>
chainctl libraries policy enable --name=team-policy --ecosystem=PYTHON --mode=ENFORCE
```

You can also combine `block` rules with a custom cooldown in the same policy:

```bash
chainctl libraries policy create --name=team-policy \
  --cooldown-days=2 \

>   --block=purl=pkg:pypi/evil

Reviewer 2: is that a real example? Maybe use some know typo/slobsquatted example instead?

Author: I added a colourama typosquat example here

chainctl libraries policy enable --name=team-policy --ecosystem=PYTHON --mode=ENFORCE
```

#### Check blocked packages

Use `chainctl libraries packages blocked` to review what a policy has blocked. By default, this shows events that have actually been enforced. For example:

```bash
chainctl libraries packages blocked --ecosystem=JAVASCRIPT
```

To see successful pulls that _would have been blocked_ by a policy that hasn't been enforced yet, use `--mode=PREVIEW`. For example:

```bash
chainctl libraries packages blocked --mode=PREVIEW --ecosystem=JAVASCRIPT
```

### Override a blocked package

Use `--allow` to permit a package that would otherwise be denied by cooldown or malware gating.

#### Override cooldown for a package version

A cooldown override is useful when a newly published version includes an urgent fix and you need it before the cooldown window expires. For example:

```bash
chainctl libraries policy update team-policy \
  --allow='purl=pkg:npm/undici@8.4.1,override-cooldown=true,justification="urgent patch"'
```

> If the new version pulls in transitive dependencies that are also still under cooldown, those transitive packages must be overridden too.

Reviewer: ```suggestion
If the new version pulls in transitive dependencies that are also blocked under a cooldown policy, those transitive packages must be overridden too.
```

chainctl libraries policy enable --policy=no-cooldown --ecosystem=JAVA --mode=ENFORCE
```

### Block a package or version

Create a custom policy with one or more `--block` entries to deny packages explicitly. Block rules are helpful if your organization has deliberately decided not to allow certain packages. To understand the impact before enforcing a block, [use `--mode=PREVIEW`](#preview-a-policy). 

For example, if your organization standardizes on React and wants to prevent teams from pulling in Angular, add a `--block` entry for each Angular package you want to deny:

```bash
chainctl libraries policy create --name=team-policy \
  --block=purl=pkg:npm/%40angular/core \
  --block=purl=pkg:npm/%40angular/common \
  --block=purl=pkg:npm/%40angular/router
chainctl libraries policy enable --name=team-policy --ecosystem=JAVASCRIPT --mode=ENFORCE
```

> To block one specific version, include the version in the purl. This is useful when a particular release is flagged with a known vulnerability that a later version has already patched, so you can deny just the affected version while allowing the other versions. For example:

Reviewer: ```suggestion
To block one specific version, include the version in the purl. For example, if a particular release is flagged with a known vulnerability that you want to block within your organization, you can block just that affected version while allowing the other versions. Use a command like the following:
```

```bash

> chainctl libraries policy create --name=team-policy \

Reviewer: ```suggestion
chainctl libraries policy update --name=team-policy \
```

Reviewer: i think we should use "update" assuming the policy has already been created, and the user wants to add things to the blocklist (now that we're not doing a full replace :) )

  --block=purl=pkg:npm/ua-parser-js@0.7.29
chainctl libraries policy enable --name=team-policy --ecosystem=JAVASCRIPT --mode=ENFORCE
```

To block all versions of a package, omit the version. For example, the following command updates the existing `team-policy` to add a block of all versions of a package:

```bash
chainctl libraries policy update team-policy \
  --block=purl=pkg:pypi/<name>
chainctl libraries policy enable --name=team-policy --ecosystem=PYTHON --mode=ENFORCE
```

You can also combine `block` rules with a custom cooldown in the same policy. The following example creates a policy that blocks `colourama`, a typosquat of the `colorama` package. In addition, a cooldown is included: 

```bash
chainctl libraries policy create --name=team-policy \
  --cooldown-days=2 \
  --block=purl=pkg:pypi/colourama
chainctl libraries policy enable --name=team-policy --ecosystem=PYTHON --mode=ENFORCE
```

#### Check blocked packages

> Use `chainctl libraries packages blocked` to review what a policy has blocked. By default, this shows events that have actually been enforced. For example:

Reviewer: ```suggestion
Use `chainctl libraries packages blocked` to review what a policy has blocked. By default, this shows pull events that were blocked any under enforced policies. For example:
```

```bash
chainctl libraries packages blocked --ecosystem=JAVASCRIPT
```

> To see successful pulls that _would have been blocked_ in the last 30 days if your current active policy had been enforced, use `--mode=PREVIEW`. For example:

Reviewer: ```suggestion
If you have policies enabled in Preview mode, you can check successful pulls that _would have been blocked_ in the last 30 days if the policy were to be enforce. Use `--mode=PREVIEW` following the example below: 
```

```bash
chainctl libraries policy enable --policy=example-policy --ecosystem=JAVASCRIPT --mode=PREVIEW
chainctl libraries packages blocked --mode=PREVIEW --ecosystem=JAVASCRIPT
```

Replace `example-policy` with the name of the policy you want to preview.

### Override a blocked package

> Use `--allow` to permit a package that would otherwise be denied by cooldown or malware gating. 

Reviewer: ```suggestion
Use `--allow` to permit a package that has been blocked by a policy, or malware and greyware blocking. 
```
